### PR TITLE
Fix #448, #594

### LINKED ISSUE #448: support for using STS to redirect plaintext to TLS
If I understood correctly, we're basically talking about integrating the functionality described here into Oragono itself: https://github.com/darwin-network/slash/issues/5

1. Designate one or more plaintext listeners as "redirect-only"
1. When clients connect to them, Oragono serves an STS policy (via `CAP LS`) plus some human-readable text about how it's necessary to reconnect via TLS

### PR DESCRIPTION
**Pull request by slingamn**
Feel free to review this or not (I might try and recruit someone else to review it), anyway I'm not in a big hurry.

### CODE REVIEW

File: irc/server.go [hunk 1]
```diff
@@ -661,39 +663,17 @@ func (server *Server) applyConfig(config *Config, initial bool) (err error) {
...
-		}
-	} else {
-		// disabling STS
-		SupportedCapabilities.Disable(caps.STS)
-		if stsPreviouslyEnabled {
-			removedCaps.Add(caps.STS)
-			stsDisabledByRehash = true
-		}
+	if (config.Server.STS.Enabled != stsPreviouslyEnabled) || (stsValue != stsCurrentCapValue) {
+		// XXX: STS is always removed by CAP NEW sts=duration=0, not CAP DEL
```
Comment (csmith) [inline irc/server.go]: Not a fan of `XXX` comments, personally. I'm left unsure whether this is a TODO that needs to be implemented or just a note about the weird behaviour.

I think in this case it's a note? It's a bit hard to follow how this works - i.e., how the cap value ends up as `duration=0` if STS is disabled. A slightly longer comment would be nice :)

File: docs/MANUAL.md [hunk 2]
```diff
@@ -342,9 +343,17 @@ If you're familiar with getting this output through your client (e.g. in weechat
 
 Otherwise, in the Oragono config file, you'll want to enable raw line logging by removing `-userinput -useroutput` under the `logging` section. Once you start up your server, connect, fail to oper and get disconnected, you'll see a bunch of input/output lines in Ora's log file. Remove your password from those logs and pass them our way.
 
+-------------------------------------------------------------------------------------------
+
+
+# IRC over TLS
+
+Traditionally, IRC used a plaintext protocol, typically on port 6667. Over time, a convention emerged to use this protocol inside SSL/TLS instead, typically on port 6697. As of now, we recommend that you make your server available *exclusively* via TLS, since allowing plaintext access can result in the disclosure of user data or passwords. While the default config file still exposes a public plaintext port for the benefit of legacy clients, it also contains instructions on how to disable it --- if at all possible, you should follow them!
```
[the remark this answers is not in the document]
Comment (DanielOaks) [reply docs/MANUAL.md]: Hmm, yeah feels a bit odd. I'd replace this paragraph with this rewrite:

>>>
IRC has traditionally been available using both plaintext (on port 6667) and SSL/TLS (on port 6697). We recommend that you make your server available exclusively via TLS, since allowing plaintext allows for unintended interception and/or disclosure of user data, communications, and passwords. While the default config file exposes a plaintext public port, it also contains instructions on how to disable it or replace it with a 'dummy' plaintext listener that simply directs users to reconnect using TLS.

File: irc/handlers.go [hunk 3]
```diff
@@ -556,6 +567,20 @@ func capHandler(server *Server, client *Client, msg ircmsg.IrcMessage, rb *Respo
 		}
 	}
 
+	sendCapLines := func(cset *caps.Set, values caps.Values) {
+		version := rb.session.capVersion
+		capLines := cset.String(version, values)
+		// weechat 1.4 has a bug here where it won't accept the CAP reply unless it contains
+		// the server.name source:
+		for i, capStr := range capLines {
+			if version == caps.Cap302 && i < len(capLines)-1 {
```
[the remark this answers is not in the document]
Comment (DanielOaks) [reply irc/handlers.go]: Yepyep, this should definitely be `>= 302`

File: irc/caps/values.go [hunk 4]
```diff
@@ -1,45 +0,0 @@
-// Copyright (c) 2017 Daniel Oaks <daniel@danieloaks.net>
-// released under the MIT license
-
-package caps
-
-import "sync"
-
-// Values holds capability values.
-type Values struct {
```
[the remark this answers is not in the document]
Comment (slingamn) [reply irc/caps/values.go]: Basically the pattern I've been going for is that all config-reloadable state is contained in the `Config` object, which is never mutated after initial load. Then the config object is accessed via atomic pointer loads and stores:

https://github.com/oragono/oragono/blob/014a5fe63c6d9ba9253318e58020274c3c6554d4/irc/getters.go#L16

so at any time, you can get a memory-safe view of the current config object.

File: irc/server.go (hunk 1 above)
Comment (slingamn) [reply irc/server.go]: It's a note. I like `XXX` a lot as a convention :-) I'll do a longer comment to go with it.
